File: src/test/java/com/automation/stepdefinitions/LoginSteps.java
package com.automation.stepdefinitions;

import com.automation.pages.LoginPage;
import com.automation.utils.ConfigReader;
import com.automation.utils.DriverFactory;
import com.automation.utils.JsonDataReader;
import com.fasterxml.jackson.databind.JsonNode;
import io.cucumber.java.en.And;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;
import org.junit.Assert;

public class LoginSteps {

    private final JsonNode loginData = JsonDataReader.readJson("login-test-data.json");
    private LoginPage loginPage;

    private LoginPage loginPage() {
        if (loginPage == null) {
            loginPage = new LoginPage(DriverFactory.getDriver());
        }
        return loginPage;
    }

    @Given("the user is on the login page")
    public void theUserIsOnTheLoginPage() {
        loginPage().open(ConfigReader.get("baseUrl"));
    }

    @When("the user enters valid username and password")
    public void theUserEntersValidUsernameAndPassword() {
        JsonNode validUser = loginData.get("validUser");
        loginPage().enterUsername(validUser.get("username").asText());
        loginPage().enterPassword(validUser.get("password").asText());
    }

    @When("the user enters invalid username and password")
    public void theUserEntersInvalidUsernameAndPassword() {
        JsonNode invalidUser = loginData.get("invalidUser");
        loginPage().enterUsername(invalidUser.get("username").asText());
        loginPage().enterPassword(invalidUser.get("password").asText());
    }

    @And("the user clicks the login button")
    public void theUserClicksTheLoginButton() {
        loginPage().clickLogin();
    }

    @Then("the user should be redirected to the dashboard")
    public void theUserShouldBeRedirectedToTheDashboard() {
        Assert.assertTrue("Dashboard is not visible after successful login", loginPage().isDashboardVisible());
    }

    @Then("the user should see a login error message")
    public void theUserShouldSeeALoginErrorMessage() {
        Assert.assertTrue("Login error is not visible for invalid credentials", loginPage().isLoginErrorVisible());
    }
}

File: src/step-definitions/login.steps.ts
import { Given, When, Then } from "@cucumber/cucumber";
import assert from "node:assert/strict";
import { CustomWorld } from "../support/world";
import { getConfig } from "../utils/config-reader";
import { readLoginData } from "../utils/test-data-reader";

const loginData = readLoginData();

Given("the user is on the login page", async function (this: CustomWorld) {
  const config = getConfig();
  await this.loginPage?.open(config.baseUrl);
});

When("the user enters valid username and password", async function (this: CustomWorld) {
  await this.loginPage?.enterUsername(loginData.validUser.username);
  await this.loginPage?.enterPassword(loginData.validUser.password);
});

When("the user enters invalid username and password", async function (this: CustomWorld) {
  await this.loginPage?.enterUsername(loginData.invalidUser.username);
  await this.loginPage?.enterPassword(loginData.invalidUser.password);
});

When("the user clicks the login button", async function (this: CustomWorld) {
  await this.loginPage?.clickLogin();
});

Then("the user should be redirected to the dashboard", async function (this: CustomWorld) {
  const visible = await this.loginPage?.isDashboardVisible();
  assert.equal(visible, true, "Dashboard should be visible after successful login");
});

Then("the user should see a login error message", async function (this: CustomWorld) {
  const visible = await this.loginPage?.isLoginErrorVisible();
  assert.equal(visible, true, "Login error should be visible for invalid credentials");
});
